src/Microsoft.Health.Core/Features/Health/AsyncCache.cs

```diff
@@ -0,0 +1,38 @@
+// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Threading.Tasks;
+using EnsureThat;
+
+namespace Microsoft.Health.Core.Features.Health;
+
+/// <summary>
+/// Asynchronously waits for the cached data to be set before the data can be retrieved
+/// </summary>
+/// <typeparam name="T">The cached data type</typeparam>
+public class AsyncCache<T> where T : class
+{
+    private T _cachedData;
+    private readonly TaskCompletionSource<bool> _initializationTask = new TaskCompletionSource<bool>();
+
+    public async Task<T> GetCachedData()
+    {
+        if (_cachedData == null)
+        {
+            await _initializationTask.Task.ConfigureAwait(false);
+        }
+
+        return _cachedData;
+    }
+
+    public void SetCachedData(T cachedData)
+    {
+        EnsureArg.IsNotNull(cachedData, nameof(cachedData));
+
+        _cachedData = cachedData;
+
+        _initializationTask.TrySetResult(true);
```

Comment:
This actually returns a `bool` result that you would need to check, but furthermore I don't think a `TaskCompletionSource` can have its result updated as that brings it to a terminal state. So unfortunately I think you have to write the class differently

Ultimately we just need an async read-writer pattern where the publisher is a single writer for multiple middleware instances potentially reading the cached report. For example, you could use a `Task<HealthReport>` as the medium, where initially you have a `TaskCompletionSource<HealthReport>.Task` for readers to read, but over time you atomically swap out that task with new ones with updated reports with like `Task.FromResult(/* my health report */)`. You probably don't even need a new class for this. The publisher itself could implement both `IHealthCheckPublisher` as well as some read-only version of `IHealthCheckReportCache` (maybe just `IHealthReportCache`)?

Reply:
Talked offline; ended up continuing with TaskCompletionSource for its simplicity and ability to await, but cleaned it up a little bit. Added CancellationToken, and only set result if the task has not been completed yet